pkg/snapshot/snapshotter/fullsnapshotleaseupdate.go

```diff
@@ -0,0 +1,72 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshotter
+
+import (
+	"context"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/health/heartbeat"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	"github.com/sirupsen/logrus"
+)
+
+// RenewFullSnapshotLeasePeriodically has a timer and will periodically call FullSnapshotCaseLeaseUpdate to renew the fullsnapshot lease until it is updated or stopped.
+// The timer starts upon snapshotter initialization and is reset after every full snapshot is taken.
+func (ssr *Snapshotter) RenewFullSnapshotLeasePeriodically() {
+	logger := logrus.NewEntry(logrus.New()).WithField("actor", "FullSnapLeaseUpdater")
+	fullSnapshotLeaseUpdateInterval := ssr.healthConfig.FullSnapshotLeaseUpdateInterval.Duration
+	ssr.FullSnapshotLeaseUpdateTimer = time.NewTimer(fullSnapshotLeaseUpdateInterval)
+	fullSnapshotLeaseUpdateCtx, fullSnapshotLeaseUpdateCancel := context.WithCancel(context.TODO())
+	defer fullSnapshotLeaseUpdateCancel()
+	logger.Infof("Starting the FullSnapshot lease renewal with interval %v", fullSnapshotLeaseUpdateInterval)
+	for {
+		select {
+		case <-ssr.FullSnapshotLeaseUpdateTimer.C:
+			if ssr.PrevFullSnapshot != nil {
+				ctx, cancel := context.WithTimeout(fullSnapshotLeaseUpdateCtx, brtypes.LeaseUpdateTimeoutDuration)
+				defer cancel()
+				if _, err := heartbeat.FullSnapshotCaseLeaseUpdate(ctx, logger, ssr.PrevFullSnapshot, ssr.K8sClientset, ssr.healthConfig.FullSnapshotLeaseName); err != nil {
+					//FullSnapshot lease update failed. Retry after interval
+					logger.Warnf("FullSnapshot lease update failed with error: %v", err)
+					logger.Infof("Resetting the FullSnapshot lease to retry updating with revision %d after %v", ssr.PrevFullSnapshot.LastRevision, fullSnapshotLeaseUpdateInterval)
+					ssr.FullSnapshotLeaseUpdateTimer.Stop()
+					ssr.FullSnapshotLeaseUpdateTimer.Reset(fullSnapshotLeaseUpdateInterval)
+				} else {
+					//FullSnapshot lease successfully updated. Stop the timer
+					logger.Infof("Stopping the FullSnapshot lease update")
+					ssr.FullSnapshotLeaseUpdateTimer.Stop()
+				}
+			} else {
+				//Skip the FullSnapshot lease update as no full snapshot has been taken yet. Reset the timer to retry after interval
+				ssr.FullSnapshotLeaseUpdateTimer.Stop()
+				ssr.FullSnapshotLeaseUpdateTimer.Reset(fullSnapshotLeaseUpdateInterval)
+			}
+
+		case <-ssr.FullSnapshotLeaseStopCh:
+			logger.Info("Closing the full snapshot lease renewal")
+			if ssr.FullSnapshotLeaseUpdateTimer != nil {
+				ssr.FullSnapshotLeaseUpdateTimer.Stop()
+				ssr.FullSnapshotLeaseUpdateTimer = nil
+			}
```

Comment:
I think having this part of the code in the ```case <-ssr.FullSnapshotLeaseStopCh:``` condition is better in terms of readability which lets the reader know for sure that the timer gets closed at this place. 

Also since there are no other places in the function where we close the timer apart from this, so putting it in defer() may imply to the reader that it gets called in several path ways. 

These are just my views, I don't have any hard opinion on this, we can change if we want WDYT?

Reply:
It's not about readability, it's about closing the resource gracefully, that's why you need to put them inside defer().

> Also since there are no other places in the function where we close the timer apart from this, so putting it in defer() may imply to the reader that it gets called in several path ways but it won't.

what makes you think so, defer inside this func, when func will get close defer will get execute.

Reply:
> what makes you think so, defer inside this func, when func will get close defer will get execute.

Sorry I didn't explain properly. But yeah having it in defer() is much safer than handling it in a if condition. I agree 👍